### ISSUE TITLE
Undefined reference to 'str' (TypeDumpVisitor.cpp.o.elf)

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-project-prepo#17]
This is spin-off from the issue SNSystems/llvm-project-prepo#7.

After fixing the SNSystems/llvm-project-prepo#10, SNSystems/llvm-project-prepo#11, SNSystems/llvm-project-prepo#12 and https://github.com/SNSystems/pstore/issues/33, running the LLVM build.

The following link error is met.
```
FAILED: : && /home/maggie/github/llvm-project-prepo/llvm/utils/repo/link.py -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -Wno-nested-anon-types  -O3 -fno-exceptions -fno-rtti  -Wl,-allow-shlib-undefined    -Wl,-rpath-link,/home/maggie/github/llvm-project-prepo/build_repo_release/./lib  -Wl,-O3 -Wl,--gc-sections tools/clang/tools/clang-check/CMakeFiles/clang-check.dir/ClangCheck.cpp.o -o bin/clang-check  lib/libLLVMX86CodeGen.a lib/libLLVMX86AsmParser.a lib/libLLVMX86AsmPrinter.a lib/libLLVMX86Desc.a lib/libLLVMX86Disassembler.a lib/libLLVMX86Info.a lib/libLLVMX86Utils.a lib/libLLVMOption.a lib/libLLVMSupport.a -lpthread lib/libclangAST.a lib/libclangBasic.a lib/libclangDriver.a lib/libclangFrontend.a lib/libclangRewriteFrontend.a lib/libclangSerialization.a lib/libclangStaticAnalyzerFrontend.a lib/libclangTooling.a lib/libLLVMAsmPrinter.a lib/libLLVMGlobalISel.a lib/libLLVMSelectionDAG.a lib/libLLVMCodeGen.a lib/libLLVMBitWriter.a lib/libLLVMScalarOpts.a lib/libLLVMAggressiveInstCombine.a lib/libLLVMInstCombine.a lib/libLLVMTransformUtils.a lib/libLLVMTarget.a lib/libLLVMAnalysis.a lib/libLLVMX86AsmPrinter.a lib/libLLVMX86Utils.a lib/libLLVMObject.a lib/libLLVMMCDisassembler.a lib/libclangStaticAnalyzerCheckers.a lib/libclangStaticAnalyzerCore.a lib/libclangCrossTU.a lib/libclangIndex.a lib/libclangFrontend.a lib/libclangDriver.a lib/libLLVMOption.a lib/libLLVMProfileData.a lib/libclangParse.a lib/libLLVMMCParser.a lib/libclangSerialization.a lib/libclangSema.a lib/libclangEdit.a lib/libclangAnalysis.a lib/libLLVMBitReader.a lib/libclangASTMatchers.a lib/libclangFormat.a lib/libclangToolingInclusions.a lib/libclangToolingCore.a lib/libclangAST.a lib/libclangRewrite.a lib/libclangLex.a lib/libclangBasic.a lib/libLLVMMC.a lib/libLLVMCore.a lib/libLLVMBinaryFormat.a lib/libLLVMDebugInfoCodeView.a lib/libLLVMDebugInfoMSF.a lib/libLLVMSupport.a -lz -ldl -ltinfo -lm lib/libLLVMDemangle.a lib/libpstore-mcrepo.a lib/libpstore-core.a -lrt lib/libpstore-broker-intf.a lib/libpstore-serialize.a lib/libpstore-support.a -lpthread -latomic -Wl,-rpath,"\$ORIGIN/../lib" && :
ERROR:wrap_tool:b"lib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0xc8): undefined reference to `.str.202'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0xd8): undefined reference to `.str.202'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x2f8): undefined reference to `.str.216'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x308): undefined reference to `.str.216'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x3e8): undefined reference to `.str.222'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x3f8): undefined reference to `.str.222'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x578): undefined reference to `.str.232'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x588): undefined reference to `.str.232'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x618): undefined reference to `.str.236'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x628): undefined reference to `.str.236'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0xa00): undefined reference to `.str.261'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0xa10): undefined reference to `.str.261'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0xaa0): undefined reference to `.str.265'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0xab0): undefined reference to `.str.265'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1ac0): undefined reference to `.str.691'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1ad0): undefined reference to `.str.691'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1ae8): undefined reference to `.str.692'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1af8): undefined reference to `.str.692'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b10): undefined reference to `.str.693'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b20): undefined reference to `.str.693'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b38): undefined reference to `.str.694'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b48): undefined reference to `.str.694'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b60): undefined reference to `.str.840'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b70): undefined reference to `.str.840'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b88): undefined reference to `.str.841'\nlib/libLLVMDebugInfoCodeView.a(TypeDumpVisitor.cpp.o.elf):
(.data.rel.ro+0x1b98): undefined reference to `.str.841'
error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by MaggieYingYi** [comment]
This issue is related with the LLVM source files.
```
1) llvm-project-prepo/llvm/lib/DebugInfo/CodeView/EnumTables.cpp.
2) llvm-project-prepo/llvm/lib/DebugInfo/CodeView/TypeDumpVisitor.cpp.
```

**Comment by MaggieYingYi** [comment]
Step 1: Compiled ‘EnumTables.cpp’ with a command something like:
```
$ /home/user/github/llvm-project-prepo/build/bin/clang++ --target=x86_64-pc-linux-gnu-repo -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Ilib/DebugInfo/CodeView -I/home/user/github/llvm-project-prepo/llvm/lib/DebugInfo/CodeView -Iinclude -I/home/user/github/llvm-project-prepo/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections  -O3 -fno-exceptions -fno-rtti    -fno-exceptions -fno-rtti -MMD -MT lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/EnumTables.cpp.o -MF lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/EnumTables.cpp.o.d -o lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/EnumTables.cpp.o -c /home/user/github/llvm-project-prepo/llvm/lib/DebugInfo/CodeView/EnumTables.cpp
```
Step 2: Get the compilation hash in object file (EnumTables.cpp.o).
```
$ repo-ticket-dump lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/EnumTables.cpp.o

a4b4b269a21751339599ed91592ff6d7
```
Step 3: Find the fragment named ‘_ZL15TrampolineNames’ in the compilation.

```
$pstore-dump -compilation= a4b4b269a21751339599ed91592ff6d7 clang.db
---

- file         : 
      path : clang.db
      size : 771751936
  compilations : 
      - digest      : a4b4b269a21751339599ed91592ff6d7
        compilation : 
            members :
                …
                - digest  : 74dbfca728c10c4be45bc78bc68b09dc
                  fext    : { addr: 233196544, size: 240 }
                  name    : _ZL15TrampolineNames
                  linkage : internal
                …
```

Step 4: Dump the fragment named ‘_ZL15TrampolineNames’ in the database.

```
$pstore-dump -fragment= 74dbfca728c10c4be45bc78bc68b09dc clang.db
---

- file      : 
      path : clang.db
      size : 771751936
  fragments : 
      - digest   : 74dbfca728c10c4be45bc78bc68b09dc
        fragment : 
            - type     : rel_ro
              contents : 
                  align   : 16
                  data    : !!binary |
                      AAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAA
                      AAAAAAwAAAAAAAAAAQAAAAAAAAA=
                  ifixups : [ ]
                  xfixups : 
                      - name   : .str.840
                        type   : 1
                        offset : 0
                        addend : 0
                      - name   : .str.840
                        type   : 1
                        offset : 16
                        addend : 0
                      - name   : .str.841
                        type   : 1
                        offset : 40
                        addend : 0
                      - name   : .str.841
                        type   : 1
                        offset : 56
                        addend : 0
```

Step 5: Print IR after RepoMetadataGeneration and Dead Global Elimination passes using -print-after-all option for the global variable ‘_ZL15TrampolineNames’.

```
*** IR Dump After RepoMetadataGenerationPass ***
@_ZL15TrampolineNames = internal global [2 x %"struct.llvm::EnumEntry.2"] zeroinitializer, align 16, !repo_ticket !844
!844 = !TicketNode(name: "_ZL15TrampolineNames", digest: [16 x i8] c"\DC\09\8B\C6\8B\C7[\E4K\0C\C1(\A7\FC\DBt", linkage: internal, pruned: false)

*** IR Dump After RepoMetadataGenerationPass ***

@.str.840 = private unnamed_addr constant [17 x i8] c"TrampIncremental\00", align 1, !repo_ticket !842
@.str.841 = private unnamed_addr constant [13 x i8] c"BranchIsland\00", align 1, !repo_ticket !843

@_ZL15TrampolineNames = internal constant [2 x %"struct.llvm::EnumEntry.2"] [%"struct.llvm::EnumEntry.2" <{ %"class.llvm::StringRef" { i8* getelementptr inbounds ([17 x i8], [17 x i8]* @.str.840, i64 0, i64 0), i64 16 }, %"class.llvm::StringRef" { i8* getelementptr inbounds ([17 x i8], [17 x i8]* @.str.840, i64 0, i64 0), i64 16 }, i16 0, [6 x i8] zeroinitializer }>, %"struct.llvm::EnumEntry.2" <{ %"class.llvm::StringRef" { i8* getelementptr inbounds ([13 x i8], [13 x i8]* @.str.841, i64 0, i64 0), i64 12 }, %"class.llvm::StringRef" { i8* getelementptr inbounds ([13 x i8], [13 x i8]* @.str.841, i64 0, i64 0), i64 12 }, i16 1, [6 x i8] zeroinitializer }>], align 16, !repo_ticket !844
!844 = !TicketNode(name: "_ZL15TrampolineNames", digest: [16 x i8] c"\DC\09\8B\C6\8B\C7[\E4K\0C\C1(\A7\FC\DBt", linkage: internal, pruned: false)
```

**Comment by MaggieYingYi** [comment]
Removed clang.db and run the above steps on the TypeDumpVisitor.cpp

```
$ rm clang.db
$ /home/user/github/llvm-project-prepo/build/bin/clang++ --target=x86_64-pc-linux-gnu-repo  -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Ilib/DebugInfo/CodeView -I/home/user/github/llvm-project-prepo/llvm/lib/DebugInfo/CodeView -Iinclude -I/home/user/github/llvm-project-prepo/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections  -O3 -fno-exceptions -fno-rtti    -fno-exceptions -fno-rtti -MMD -MT lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/TypeDumpVisitor.cpp.o -MF lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/TypeDumpVisitor.cpp.o.d -o lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/TypeDumpVisitor.cpp.o -c /home/user/github/llvm-project-prepo/llvm/lib/DebugInfo/CodeView/TypeDumpVisitor.cpp
$/home/user/github/llvm-project-prepo/build/bin/repo-ticket-dump lib/DebugInfo/CodeView/CMakeFiles/LLVMDebugInfoCodeView.dir/TypeDumpVisitor.cpp.o
$/home/user/github/llvm-project-prepo/build/bin/pstore-dump -compilation= d72bf2b8e47f0446f996b277c2de1f5f clang.db
$ /home/user/github/llvm-project-prepo/build/bin/pstore-dump -fragment= 74dbfca728c10c4be45bc78bc68b09dc clang.db
```
We could see that the global variables ‘_ZL15TrampolineNames’ and ‘_ZL13LabelTypeEnum’ have the same fragment .

Print IR after RepoMetadataGeneration and Dead Global Elimination passes for the global variable ‘_ZL13LabelTypeEnum’ in the TypeDumpVisitor.cpp.



*** IR Dump After RepoMetadataGenerationPass ***

@_ZL13LabelTypeEnum = internal global [2 x %"struct.llvm::EnumEntry.0"] zeroinitializer, align 16, !repo_ticket !267
!267 = !TicketNode(name: "_ZL13LabelTypeEnum", digest: [16 x i8] c"\DC\09\8B\C6\8B\C7[\E4K\0C\C1(\A7\FC\DBt", linkage: internal, pruned: false)


*** IR Dump After Dead Global Elimination ***
@.str.266 = private unnamed_addr constant [5 x i8] c"Near\00", align 1, !repo_ticket !268
@.str.267 = private unnamed_addr constant [4 x i8] c"Far\00", align 1, !repo_ticket !269
@_ZL13LabelTypeEnum = internal constant [2 x %"struct.llvm::EnumEntry.0"] [%"struct.llvm::EnumEntry.0" <{ %"class.llvm::StringRef" { i8* getelementptr inbounds ([5 x i8], [5 x i8]* @.str.266, i64 0, i64 0), i64 4 }, %"class.llvm::StringRef" { i8* getelementptr inbounds ([5 x i8], [5 x i8]* @.str.266, i64 0, i64 0), i64 4 }, i16 0, [6 x i8] zeroinitializer }>, %"struct.llvm::EnumEntry.0" <{ %"class.llvm::StringRef" { i8* getelementptr inbounds ([4 x i8], [4 x i8]* @.str.267, i64 0, i64 0), i64 3 }, %"class.llvm::StringRef" { i8* getelementptr inbounds ([4 x i8], [4 x i8]* @.str.267, i64 0, i64 0), i64 3 }, i16 4, [6 x i8] zeroinitializer }>], align 16, !repo_ticket !267

!267 = !TicketNode(name: "_ZL13LabelTypeEnum", digest: [16 x i8] c"\DC\09\8B\C6\8B\C7[\E4K\0C\C1(\A7\FC\DBt", linkage: internal, pruned: false)
```

**Comment by MaggieYingYi** [comment]
The three small test cases are used to reproduce the problem.

The first IR test code (test.ll):
```
$ less test.ll

target triple = "x86_64-pc-linux-gnu-repo"

declare i32 @printf(i8*, ...)

@.str = private unnamed_addr constant [13 x i8] c"Hello World\0A\00", align 1

define i32 @foo() !repo_ticket !0 {
entry:
  %call = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([13 x i8], [13 x i8]* @.str, i32 0, i32 0))
  ret i32 0
}

!repo.tickets = !{!0}

!0 = !TicketNode(name: "foo", digest: [16 x i8] c"N\DB\88\CF\F2a\8D.\8C!yNR[|d", linkage: external, pruned: false)

```
The second IR test code (test1.ll):
```
$ less test1.ll

target triple = "x86_64-pc-linux-gnu-repo"

declare i32 @printf(i8*, ...)

@.str = private unnamed_addr constant [13 x i8] c"MyFirstTest\0A\00", align 1

define i32 @bar() !repo_ticket !0 {
entry:
  %call = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([13 x i8], [13 x i8]* @.str, i32 0, i32 0))
  ret i32 0
}

!repo.tickets = !{!0}

!0 = !TicketNode(name: "bar", digest: [16 x i8] c"N\DB\88\CF\F2a\8D.\8C!yNR[|d", linkage: external, pruned: false)
```

The third IR test code (test2.ll):
```
$ less test2.ll

target triple = "x86_64-pc-linux-gnu-repo"

declare i32 @printf(i8*, ...)

@.str.1 = private unnamed_addr constant [13 x i8] c"Hello World\0A\00", align 1

define i32 @quz() !repo_ticket !0 {
entry:
  %call = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([13 x i8], [13 x i8]* @.str.1, i32 0, i32 0))
  ret i32 0
}

!repo.tickets = !{!0}

!0 = !TicketNode(name: "quz", digest: [16 x i8] c"N\DB\88\CF\F2a\8D.\8C!yNR[|d", linkage: external, pruned: false)

```

Assuming:
1) Global functions 'foo', 'bar' and 'quz' have the same digest during the RepoMetadataGeneration pass.
2) .str and .str.1 are generated during optimization and not having the TicketNode.

**Comment by MaggieYingYi** [comment]
Compiling above tests.

```
$ rm clang.db
$ llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj test.ll -o test.o
$ llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj test1.ll -o test1.o
$ llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj test2.ll -o test2.o
$ pstore-dump -all-compilations  clang.db
---

- file         : 
      path : clang.db
      size : 4194304
  compilations : 
      - digest      : 506d090d8ad114093e4929731a4ca6ef
        compilation : 
            members : 
                - digest  : 647c5b524e79218c2e8d61f2cf88db4e
                  fext    : { addr: 968, size: 144 }
                  name    : bar
                  linkage : external
                - digest  : 5f32287ed3b9a45334776815cf836f4f
                  fext    : { addr: 1112, size: 45 }
                  name    : .str
                  linkage : internal
            path    : \\\\?\\C:\\MyWork\\repo_bug\\issue17
            triple  : x86_64-pc-linux-gnu-repo
      - digest      : 5772cb11c7f59fd3c79639667074df14
        compilation : 
            members : 
                - digest  : 647c5b524e79218c2e8d61f2cf88db4e
                  fext    : { addr: 352, size: 144 }
                  name    : foo
                  linkage : external
                - digest  : e5ff1cfe5481aaf0305b2a111a14012e
                  fext    : { addr: 272, size: 45 }
                  name    : .str
                  linkage : internal
            path    : \\\\?\\C:\\MyWork\\repo_bug\\issue17
            triple  : x86_64-pc-linux-gnu-repo
      - digest      : b714184e7f07b0f73ff52cca7031c24d
        compilation : 
            members : 
                - digest  : 647c5b524e79218c2e8d61f2cf88db4e
                  fext    : { addr: 1752, size: 144 }
                  name    : quz
                  linkage : external
                - digest  : e5ff1cfe5481aaf0305b2a111a14012e
                  fext    : { addr: 1704, size: 45 }
                  name    : .str.1
                  linkage : internal
            path    : \\\\?\\C:\\MyWork\\repo_bug\\issue17
            triple  : x86_64-pc-linux-gnu-repo
...
```

There are two issues here.
```
1) Currently, the functions 'foo' and 'bar' have the same digest. It is wrong.
   Both 'foo' and 'bar' have a dependent '.str', but the content (or digest) of '.str' 
   in the test.ll and test1.ll are different. Therefore, functions 'foo' and 'bar' can't
   share the same fragment.
   
2) Currently, the functions 'foo' and 'quz' have the same digest. It is wrong.
   The function 'foo' has a dependent '.str' and the function 'bar' has a dependent '.str.1'. 
   The content (or digest) of '.str' and '.str.1' in the test.ll and test2.ll are the same.
   If functions 'foo' and 'bar' share the same fragment, the local variable '.str' can only
   be seen by the 'test.o' compilation, but not visible to the ‘test1.o’ compilation. This 
   caused the link error: undefined reference to ‘.str’.
```

Therefore, the functions 'foo', 'bar' and 'quz' shoudn't share the same fragment.

**Comment by paulhuggett** [comment]
> Currently, the functions 'foo' and 'bar' have the same digest. It is wrong. Both 'foo' and 'bar' have a dependent '.str', but the content (or digest) of '.str' in the test.ll and test1.ll are different. Therefore, functions 'foo' and 'bar' can't share the same fragment.

I don't think there's a problem with sharing a fragment between two functions which refer to an identically named static variable. Consider:

    –a.c-
    static int v = 1;
    int f (void) {
        return v;
    }

    –b.c-
    static int v = 2;
    int g (void) {
        return v;
    }

The two functions here (`f` and `g`) have identical bodies and both refer to (different) static variables named `v`. The fragments for `f` and `g` can be the same because their code and i/fixups are identical; the fragments for `a.c:v` and `b.c:v` must be different (because the initial values of the two globals are different). The following pseudo pstore-dump output shows this (here I've used `d(h)` to denote the fragment shared by `f` and `g`).

    - file: 
          path : clang.db
          size : 1
      fragments : 
          - digest   : d(h)
            fragment : 
                - type     : text
                  contents : 
                      data    : ...
                      ifixups : [ ]
                      xfixups : [{ name: v, type: 1, offset: 0, addend: 0 }]
          - digest   : d(a.c:v)
            fragment : { type: data, contents: "int 1" }
          - digest   : d(b.c:v)
            fragment : { type: data, contents: "int 2" }
      compilations : 
          - digest      : digest(a.c)
            compilation : 
                members : 
                    - { digest: d(h),     name: f, linkage: external }
                    - { digest: d(a.c:v), name: v, linkage: internal }
          - digest      : digest(b.c)
            compilation : 
                members : 
                    - { digest: d(h),     name: g, linkage: external }
                    - { digest: d(b.c:v), name: v, linkage: internal }

Does this make sense?

**Comment by paulhuggett** [comment]
> Currently, the functions 'foo' and 'quz' have the same digest. It is wrong. The function 'foo' has a dependent '.str' and the function 'bar' has a dependent '.str.1'.

Yes, absolutely. The fragments for your `foo` and `quz` functions will each contain a xfixup referring to `.str` and `.str.1` respectively. These named must be considered when generating the digests for the two functions. To reduce this to a minimal pair of .c files:

    –a.c-
    static int x = 1;
    int f () {
        return x;
    }
    –b.c-
    static int y = 1;
    int g () {
        return y;
    }

Clearly here the fragments for `f()` and `g()` must be different because there will be an xfixup containing the name of the variable being referenced (`x` and `y` respectively).

**Comment by MaggieYingYi** [comment]
>  
> > Currently, the functions 'foo' and 'bar' have the same digest. It is wrong. Both 'foo' and 'bar' have a dependent '.str', but the content (or digest) of '.str' in the test.ll and test1.ll are different. Therefore, functions 'foo' and 'bar' can't share the same fragment.
>
> I don't think there's a problem with sharing a fragment between two functions which refer to an identically named static variable. 
>
> Does this make sense?

Depending on whether the compiler inlines the ‘v’ or not. If the ‘v’ is inlined, the functions ‘f’ and ‘g’ have the different digest. The inlining issue has been considered in the current implementation of the llvm-prepo project. If you build the repo with HEAD of master branch (without my patch) and compiled the a.c and b.c using the built llvm-prepo compiler, you will find that the functions ‘f’ and ‘g’ have different digests. Since we don’t know the ‘v’ will be inlined or not when generating the digest of function ‘f’, we have to accumulate the digest of ‘v’ to the digest of ‘f’ and ‘g’. 

This issue is a different case. The ‘.str’ exists not only in xfixup but also in the dependent.  
Considering the test.ll, the ‘.str’ is generated by a LLVM optimisation pass ( it’s after the RepoMetadataGeneration pass), the ‘.str’ doesn’t have a TiketNode metadata which will be generated by the backend (llc). 
```
declare i32 @printf(i8*, ...)
@.str = private unnamed_addr constant [13 x i8] c"Hello World\0A\00", align 1
define i32 @foo() !repo_ticket !0 {
entry:
  %call = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([13 x i8], [13 x i8]* @.str, i32 0, i32 0))
  ret i32 0
}
!repo.tickets = !{!0}
!0 = !TicketNode(name: "foo", digest: [16 x i8] c"N\DB\88\CF\F2a\8D.\8C!yNR[|d", linkage: external, pruned: false)
```

Dumping the fragment of function ‘foo’
```
- file      :
      path : clang.db
      size : 4194304
  fragments :
      - digest   : bb3ad6a9eebab0e7b41cc605487a0da8
        fragment :
            - type     : text
              contents :
                  align   : 16
                  data    :
                      - 0:      pushq   %rax
                      - 1:      movl    $0, %edi
                      - 6:      xorl    %eax, %eax
                      - 8:      callq   0
                      - 13:     xorl    %eax, %eax
                      - 15:     popq    %rcx
                      - 16:     retq
                  ifixups : [ ]
                  xfixups :
                      - name   : .str                         <- ‘.str’ is a xfixup.
                        type   : 10
                        offset : 2
                        addend : 0
                      - name   : printf
                        type   : 4
                        offset : 9
                        addend : 18446744073709551612
            - type     : dependent
              contents :
                  - digest  : e5ff1cfe5481aaf0305b2a111a14012e
                    fext    : { addr: 280, size: 45 }
                    name    : .str                             <- ‘.str’ is a dependent.
                    linkage : internal
```
The digest of the fragment 'foo' didn’t consider its dependents (the digest is generated by backend and exists in the fragment as dependent type). The proposed patch (https://github.com/SNSystems/llvm-project-prepo/pull/18) try to solve this issue.

**Comment by paulhuggett** [comment]
> If the ‘v’ is inlined, the functions ‘f’ and ‘g’ have the different digest.

Can you show an example of what you mean by 'v' being "inlined"? Is there anything that the compiler can do beyond constant-propagation if the variable is never assigned?

> This issue is a different case. The ‘.str’ exists not only in xfixup but also in the dependent.  

Ah, okay. This is an example of the compiler synthesising a new static variable. Our classic example is a call to `printf()` with a string containing no formatting and ending with `\n`:

    static char const * str = "hello\n";
    void h () { printf (str); }

An optimization pass later turns this into:

    static char const * str1 = "hello";
    void h () { puts (str1); }

The hash-generation pass didn't see `str1` so it falls to the back-end to generate its hash and record that `h()` is dependent on `str1` ("dependence" in this case is recorded in the fragment and tells a . (I'm aware that I'm repeating what you wrote above, but with examples in  C rather than IR because it's more compact and easier [for me] to follow!)

> The digest of the fragment 'foo' didn’t consider its dependents (the digest is generated by backend and exists in the fragment as dependent type).

Yes, this makes sense. We place a link in fragment's "dependent" section to tell an incremental compile that materializes this fragment from the database that it must also create a compilation member for `str1`.

Since the name and digest of the dependee appears in the fragment it follows that we must include this information in the fragment's digest.

**Comment by MaggieYingYi** [comment]
Thanks for reproducing the issue in C.

> 
> 
> > If the ‘v’ is inlined, the functions ‘f’ and ‘g’ have the different digest.
> 
> Can you show an example of what you mean by 'v' being "inlined"? 
>

Compiling the above small code a.c and b.c with -O3.
```
$ rm clang.db
$ clang -O3 -c --target="x86_64-pc-linux-gnu-repo"  a.c -o a.o
$ clang -O3 -c --target="x86_64-pc-linux-gnu-repo"  b.c -o b.o
$ pstore-dump -all-fragments clang.db
- file      :
      path : clang.db
      size : 4194304
  fragments :
      - digest   : e13406d828bab627123856a2986ce8a2
        fragment :
            - type     : text
              contents :
                  align   : 16
                  data    :
                      - 0:      movl    $1, %eax
                      - 5:      retq
                  ifixups : [ ]
                  xfixups : [ ]
      - digest   : 1a437acfbd77a678773fe23efee6f4f9
        fragment :
            - type     : text
              contents :
                  align   : 16
                  data    :
                      - 0:      movl    $2, %eax
                      - 5:      retq
                  ifixups : [ ]
                  xfixups : [ ]
...
```

The digests of functions 'f' and 'g' are different because the 'v' is inlined into both functions 'f' and 'g'.

> Is there anything that the compiler can do beyond constant-propagation if the variable is never assigned?
> 
I am not sure, but If the variable is not used, the compiler could remove it using dead code elimination optimization.

**Comment by MaggieYingYi** [comment]
The fix has been committed to the master branch.

**Comment by MaggieYingYi** [comment]
As mentioned in the https://github.com/SNSystems/llvm-project-prepo/issues/38#issuecomment-525196748, the commit 65d8c8d has fixed the issue which is described in the beginning. 

The previous fix for this issue (dd5de96e828) need to be reverted. The changes are currently under the code review https://github.com/SNSystems/llvm-project-prepo/pull/39. 

I have discussed the second issue with Paul. He will create a new issue for the second one.
